### ISSUE TITLE
Ionic View > Close (&times;) button not working

### ISSUE THREAD

**Issue by martinpritchardelevate** [filestack/filestack-js#92]
```
// package.json snippet...
...
"filestack-js": "0.11.2",
"@angular/core": "5.0.0",
ionic-angular: "3.9.2",
typescript: "2.4.2",
...
```

I'm using filestack in an ionic 3 app, and when running via the 'ionic view' or via cordova, app the close/&times; button does not work....which means users are stuck in the filestack modal. This appears to only be a problem on andriod devices.

**Comment by schmich** [comment]
I hit this as well when using the Filestack JS API (v0.11.5) in a PhoneGap/Cordova app running on Android. This is a blocking issue that prevents the use of Filestack entirely on Cordova, at least on Android (I haven't checked iOS).

I found that the issue stems from a conflict between Filestack and Cordova's network connectivity handling.

This is a roughly how we're invoking the picker:

```js
async function pickFile() {
  await loadScript('https://static.filestackapi.com/v3/filestack.js');
  const apiKey = 'xxx';
  const client = filestack.init(apiKey);
  const options = {
    accept: 'image/*',
    minFiles: 1,
    maxFiles: 1,
    uploadInBackground: false
  };
  await client.pick(options);
}
```

When the user attempts to close the modal _or_ even when they successfully upload an image, the library crashes when trying to cleanup the modal:

```
Uncaught (in promise) TypeError: Cannot read property 'handleEvent' of undefined
    at checkSubscriptionArgument (<anonymous>:742:59)
    at Channel.unsubscribe (<anonymous>:801:5)
    at window.removeEventListener (<anonymous>:158:32)
    at Us.removeNetworkListeners (network.js:51)
    at Array.<anonymous> (vuex.esm.js:704)
    at Us.dispatch (vuex.esm.js:426)
    at Object.Us.dispatch (vuex.esm.js:332)
    at Us.allUploadsDone (store.js:128)
    at Array.<anonymous> (vuex.esm.js:704)
    at Us.dispatch (vuex.esm.js:426)
    at Us.dispatch (vuex.esm.js:332)
    at Us.finishUploadsMaybe (upload_queue.js:459)
    at Array.<anonymous> (vuex.esm.js:704)
    at Us.dispatch (vuex.esm.js:426)
    at Object.Us.dispatch (vuex.esm.js:332)
    at Us.uploadMoreMaybe (upload_queue.js:473)
    at Array.<anonymous> (vuex.esm.js:704)
    at Us.dispatch (vuex.esm.js:426)
    at Object.Us.dispatch (vuex.esm.js:332)
    at upload_queue.js:470
```

Digging through the Filestack calls on the stack, it looks like Filestack is trying to cleanup its network connectivity event listeners:

```js
// at Us.allUploadsDone (store.js:128)
context.dispatch('removeNetworkListeners'); // <<< callsite
```

```js
// at Us.removeNetworkListeners (network.js:51)
removeNetworkListeners({ state: { listeners: { online, offline } } }) {
  window.removeEventListener('online', online); // <<< callsite
  window.removeEventListener('offline', offline);
}
```

This is where the issue starts. Normally, the call to `window.removeEventListener` would be uninteresting and would succeed. However, Cordova actually shims and intercepts calls to `window.removeEventListener`:

```js
// at window.removeEventListener (<anonymous>:158:32)
window.removeEventListener = function (evt, handler, capture) {
    var e = evt.toLowerCase();
    // If unsubscribing from an event that is handled by a plugin
    if (typeof windowEventHandlers[e] !== 'undefined') {
        windowEventHandlers[e].unsubscribe(handler); // <<< callsite
    } else {
        m_window_removeEventListener.call(window, evt, handler, capture);
    }
};
```

The `online` and `offline` events on `window` are specially handled by Cordova, and digging deeper, the error occurs because Cordova expects the event handlers to conform to a specific interface.

Digging through the Android Cordova source (seen at [apache/cordova-android/cordova-js-src/exec.js](https://github.com/apache/cordova-android/blob/ad01d28351c13390aff4549258a0f06882df59f5/cordova-js-src/exec.js)), Cordova does some redirection magic to bridge `online`/`offline` events from `document` to `window` in conjunction with its [network information](https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-network-information/index.html) plugin:

```js
function hookOnlineApis() {
    function proxyEvent(e) {
        cordova.fireWindowEvent(e.type);
    }
    // The network module takes care of firing online and offline events.
    // It currently fires them only on document though, so we bridge them
    // to window here (while first listening for exec()-releated online/offline
    // events).
    window.addEventListener('online', pollOnceFromOnlineEvent, false);
    window.addEventListener('offline', pollOnceFromOnlineEvent, false);
    cordova.addWindowEventHandler('online');
    cordova.addWindowEventHandler('offline');
    document.addEventListener('online', proxyEvent, false);
    document.addEventListener('offline', proxyEvent, false);
}
```

I haven't investigated further to figure out why they're doing this or what purpose it serves.

The correct long-term fix might be Cordova ignoring non-conforming event handlers, but a potential workaround on the Filestack side could be to just listen for the `online`/`offline` events on `document.body` instead of `window`. This should work according to [MDN](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/Online_and_offline_events) and the HTML spec (see [under section 8.1.5.2](https://html.spec.whatwg.org/multipage/webappapis.html#event-handlers-on-elements,-document-objects,-and-window-objects)):

> The following are the event handlers (and their corresponding event handler event types) that must be supported by Window objects, as event handler IDL attributes on the Window objects themselves, and with corresponding event handler content attributes and event handler IDL attributes exposed on all body and frameset elements that are owned by that Window object's associated Document

Also, interestingly, on the Filestack side, the `online` and `offline` event handlers being removed in `removeNetworkListeners` are actually both `undefined`. In other words, in my debugging, Filestack never actually handles the connectivity events, but it still crashes when trying to unsubscribe due to the conflict with Cordova.

Thoughts?

**Comment by schmich** [comment]
A _terrible, terrible_ workaround that _nobody_ in their right mind should use:

```js
async function hookWindowEvents(fn) {
  const realRemove = window.removeEventListener;
  const hookRemove = function(name, handler) {
    if (handler) {
      return realRemove.apply(window, arguments);
    }
  };

  try {
    window.removeEventListener = hookRemove;
    return await fn();
  } finally {
    window.removeEventListener = realRemove;
  }
}

async function pickFile() {
  await loadScript('https://static.filestackapi.com/v3/filestack.js');
  const apiKey = 'xxx';
  const client = filestack.init(apiKey);
  const options = { accept: 'image/*' };
  return await hookWindowEvents(async () => await client.pick(options));
}
```

This just intercepts and ignores any calls to `window.removeEventListener` without a handler for the duration of the picker. This works in the narrow case I'm testing with, but will fail if the Filestack API decides to handle `online`/`offline` events with real handlers.

**Comment by velveteer** [comment]
@schmich Now this is a curious case, but we tend to see such things when dealing with globals. :stuck_out_tongue_closed_eyes: 
 
I believe the option `allowManualRetry` is why these network listeners are attached, but it appears that even when this option is false we attempt to remove listeners that do not exist. This is a really good find and is something I think we can address, likely by not removing the handlers if we don't need to, and also looking into using `document.body`.

**Comment by schmich** [comment]
That would be great to see. If you need a minimal repro for Cordova + Android, let me know and I'll put one together.

**Comment by velveteer** [comment]
@schmich It turns out that offline/online events on the document are against the spec, as discussed here: https://bugzilla.mozilla.org/show_bug.cgi?id=1457166. And I could not get the logic we need for `allowManualRetry` to work using event handlers on the document.

However, we did fix the issue where `window.removeEventListener` was being called without reason. So now the presence of `allowManualRetry: true` controls whether or not these handlers are attached and subsequently removed on close. It's possible that this option may be incompatible with Cordova due to our reliance on window events. This patch was published today in 1.0.2.

I see that you are on a pre-1.0 version of this library, but if you do migrate to 1.0.2 it would be great if you  could confirm the patch resolves this particular issue.

**Comment by schmich** [comment]
@velveteer Thanks! This works for our scenarios. I appreciate the effort and can understand that it won't necessarily work when the event handlers are required.

As far as I'm concerned, this bug seems as fixed as it's going to be, so I imagine you can close this issue.
